### PR TITLE
TIP-0003: Subset Argument for Memory Consistency

### PR DESCRIPTION
**Pull request by aszepieniec**
This tip proposes an AIR for establishing the correct sorting within each contiguous region of memory-like tables. It is an alternative to TIP-0002 and together with TIP-0001 it fixes the weakness identified by Yuncong's attack.

### CODE REVIEW

**Review by jan-ferdinand — CHANGES_REQUESTED** [review]
For some reason, I find TIP 0003 more difficult to digest than your usual writings. It might have to do with the intuition being not as intuitive or complete as it could be. How about adding a paragraph containing the following line of thought?

“Conceptually, the presented argument is quite similar to the Instruction Table: an Evaluation Argument establishes that only correct values can be used, a Permutation Argument establishes that those values can be used as often as need be. In the Instruction Table, the “value” to be used is the combination of (address, instruction), in the presented argument, it is clock jump differences.”

Possibly, a (cleaned-up) version of this picture might help. However, there is a point at which we over-engineer the TIP.

File: tips/tip-0003/tip-0003.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+Unfortunately TIP-0002 introduces a lot of design complexity by. The cause of this complexity is the Polynomial IOP for correct calculation of the formal derivative.
+
+This TIP proposes an alternative way to establish the correct calculation of the formal derivative, namely through a more traditional AIR/RAP construction. While there is a performance penalty of one extra polynomial commitment, the end-result is a noticeably simpler design.
+
+TIPs 0002 and 0003 are mutually exclusive companions to TIP-0001, which introduces an new argument to establish the contiguity of regions of constant memory pointer. Together, TIP-0001 and (TIP-0002 or TIP-0003) fix the memory consistency issue.
+
+## Introduction
+
+How to establish that a clock jump is directed forward (as opposed to direct backward, which would indicate malicious behavior)? One strategy is to show that the *difference*, *i.e.*, the next clock cycle minus the current clock cycle, is itself a clock cycle. Recall that the Processor Table's clock cycles run from 0 to $T-1$. Therefore, valid differences belong to $\lbrace 1, ..., T-1 \rbrace$ and invalid ones to $\lbrace p - T + 1, ..., p - 1 \rbrace$.
```
[the remark this answers is not in the document]
Comment (aszepieniec) [reply tips/tip-0003/tip-0003.md:22]: Yes. One of the conclusions of TIP-0002 was that ignoring clock jump differences of one is very hard, so now they are included.

File: tips/tip-0003/tip-0003.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+### Clock Jump Differences with Multiplicities
+
+The clock jump differences of the memory-like tables are all listed in the $\mathsf{cjd}$ column of the Processor Table. The values are sorted and the padding inserts "0" rows at the bottom. This relation comes with another extension column computing a running product. Denote this column by $\mathsf{rp}$.
+
+The initial value of this column is the product of the initial values of the matching columns of the Ram Table, JumpStack Table, and OpStack Table. This gives rise to the following cross-table initial boundary constraint:
+
+$$ \mathsf{rp} - \mathsf{rp}^{(r)} \cdot \mathsf{rp}^{(j)} \cdot \mathsf{rp}^{(o)} $$
+
+This running product accumulates a factor $(X - \mathsf{cjd})$ in every row where $\mathsf{cjd} \neq 0$. Assume there is a column $\mathsf{invm}$ which is which is the inverse-or-zero of $\mathsf{cjd}$. Then the transition constraint is
```
[the remark this answers is not in the document]
Comment (aszepieniec) [reply tips/tip-0003/tip-0003.md]: Commit [9f761d7](https://github.com/TritonVM/triton-vm/pull/69/commits/9f761d70dd5a7ebeec6949164594f5bfb20dcdbc) supplies an etymology along the lines of *clock jump differences with mulitplicities*.

File: tips/tip-0003/tip-0003.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ - In addition to the extension column computing the running product, the Processor Table needs an inverse column $\mathsf{invm}$ to help select all *nonzero* $\mathsf{cjd}$'s, and thus skip padding rows. The abbreviation *invm* is short for inverse-with-multiplicities.
+ - An inverse column $\mathsf{invu}$ in the Processor Table allows for selecting the last row of *every* contiguous region of $\mathsf{cjd}$. The abbreviation *invu* is short for unique-inverse.
+ - An evaluation argument establishes that a selection of clock cycles and the selected clock jump differences are identical lists. This evaluation argument requires two more extension columns on the Processor Table: $\mathsf{rer}$ computes the running evaluation for the *relevant* clock cycles, whereas $\mathsf{reu}$ computes the running evaluation for the *unique* clock jump differences.
+
+The total is 3 extra base columns and 6 extension columns.
+
+![](relations-diagram.svg)
```
Comment (jan-ferdinand) [inline tips/tip-0003/tip-0003.md:41]: That diagram is gorgeous!

File: tips/tip-0003/tip-0003.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+This TIP proposes an alternative way to establish the requisite set relation, namely through a more traditional AIR/RAP construction that avoids computing the formal derivative entirely. While there is a performance penalty of two extra polynomial commitments, the end-result is a noticeably simpler design.
+
+TIPs 0002 and 0003 are mutually exclusive companions to TIP-0001, which introduces an new argument to establish the contiguity of regions of constant memory pointer. Together, TIP-0001 and (TIP-0002 or TIP-0003) fix the memory consistency issue.
+
+## Introduction
+
+How to establish that a clock jump is directed forward (as opposed to direct backward, which would indicate malicious behavior)? One strategy is to show that the *difference*, *i.e.*, the next clock cycle minus the current clock cycle, is itself a clock cycle. Recall that the Processor Table's clock cycles run from 0 to $T-1$. Therefore, valid differences belong to $\lbrace 1, ..., T-1 \rbrace$ and invalid ones to $\lbrace p - T + 1, ..., p - 1 \rbrace$.
```
Comment (jan-ferdinand) [inline tips/tip-0003/tip-0003.md:22]: Is it worth pointing out that $T$ can never be so large as to be a problem for very far backwards jumps?

File: tips/tip-0003/tip-0003.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+The total is 3 extra base columns and 6 extension columns.
+
+![](relations-diagram.svg)
+
+## Detailed Description
+
+### Memory-like Tables
+
+Here are the constraints for the RAM Table. The constraints for other two tables are analogous and are therefore omitted from this section. Elsewhere the superscripts $^{(r)}$, $^{(j)}$ and $^{(o)}$ disambiguate between the RAM Table, JumpStack Table, and OpStack Table, respectively. The constraints are relative to the indeterminate $X$ which anticipates the verifier's challenge $\alpha$ via the implicit substitution $X \mapsto \alpha$.
```
Comment (jan-ferdinand) [inline tips/tip-0003/tip-0003.md:47]:
```suggestion
Here are the constraints for the RAM Table. The constraints for the other two tables are analogous and are therefore omitted from this section. Elsewhere, the superscripts $^{(r)}$, $^{(j)}$ and $^{(o)}$ disambiguate between the RAM Table, JumpStack Table, and OpStack Table, respectively. The constraints are relative to the indeterminate $X$ which anticipates the verifier's challenge $\alpha$ via the implicit substitution $X \mapsto \alpha$.
```

**Review by jan-ferdinand — APPROVED** [review]

**Pull request merged**